app/src/main/java/org/worldbank/transport/driver/activities/RecordFormActivity.java

```diff
@@ -40,51 +45,334 @@
+    public static final String SECTION_ID = "driver_section_id";
+    private static final String LOG_LABEL = "RecordFormActivity";
+
+    private DriverAppContext mAppContext;
+    DriverApp app;
+
+    // flag that is true once form has been displayed
+    private boolean formReady = false;
+    private FormReadyListener formReadyListener;
+
+    private DriverSchema currentlyEditing;
+    private int sectionId;
+    private String sectionName;
+    private String[] sectionOrder;
+    private Class sectionClass;
+    private String sectionLabel;
+
+    // have next/previous sections
+    private boolean haveNext = false;
+    private boolean havePrevious = false;
+
+    // if selected action is to go to previous (if false, go to next or save)
+    private boolean goPrevious = false;
+
+    /**
+     * Non-default constructor for testing, to set the application context.
+     * @param context Mock context
+     */
+    public RecordFormActivity(DriverAppContext context) {
+        super();
+        mAppContext = context;
+    }
+
+    /**
+     * Default constructor, for testing.
+     */
+    public RecordFormActivity() {
+        super();
+    }
+
     @Override
-    public void onCreate(Bundle savedInstanceState) {
+    protected void onCreate(Bundle savedInstanceState) {
+        // set up some state before calling super
+        mAppContext = new DriverAppContext((DriverApp) getApplicationContext());
+        app = mAppContext.getDriverApp();
+        currentlyEditing = app.getEditObject();
+
+        Bundle bundle = getIntent().getExtras();
+        sectionId = bundle.getInt(SECTION_ID);
+        sectionOrder = app.getSchemaSectionOrder();
+
+        // calling super will call createFormController in turn
         super.onCreate(savedInstanceState);
+    }
+
+    /**
+     * This method gets called by the form building background task after it finishes.
+     */
+    @Override
+    public void displayForm() {
+        super.displayForm();
 
+        // now form has been built, add next or save button to it
         ViewGroup containerView = (ViewGroup) findViewById(R.id.form_elements_container);
 
+        // reference to this, for use in button actions (validation task makes weak ref)
+        final RecordFormActivity thisActivity = this;
+
+        // put buttons in a relative layout for positioning on right or left
+        RelativeLayout buttonBar = new RelativeLayout(this);
+        buttonBar.setId(R.id.record_button_bar_id);
+        buttonBar.setLayoutParams(new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.MATCH_PARENT,
+                RelativeLayout.LayoutParams.WRAP_CONTENT));
+
+        // add 'previous' button
+        if (sectionId > 0) {
+            havePrevious = true;
+            Button backBtn = new Button(this);
+            RelativeLayout.LayoutParams backBtnLayoutParams = new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT,
```

Comment:
It might be good to split some of this button-adding code out into a function or functions; I'm not sure how much code reuse you'll be able to get out of it but it might make the `displayForm` function easier to grok since it would add a clear way to break it up mentally.